packages/teams-js/src/internal/validDomains.ts

```diff
@@ -0,0 +1,104 @@
+import { GlobalVars } from '../internal/globalVars';
+import { validDomainsCdnEndpoint, validOriginsFallback } from './constants';
+import { getLogger } from './telemetry';
+import { isValidHttpsURL } from './utils';
+
+let validOrigins: string[] = [];
+export async function prefetchDomainsFromCDN(): Promise<void> {
+  if (fetch) {
+    await fetch(validDomainsCdnEndpoint);
+  }
+}
+
+async function retrieveDomainsFromCDNAndStore(): Promise<string[]> {
+  if (validOrigins.length !== 0) {
+    return validOrigins;
+  }
+  if (fetch) {
+    return fetch(validDomainsCdnEndpoint)
+      .then((response) => {
+        if (!response.ok) {
+          validOrigins = validOriginsFallback;
+          return validOriginsFallback;
+        }
+        return response.json().then((validDomains) => {
+          validOrigins = validDomains.validOrigins;
```

Comment:
Should the data retrieved be validated? To make sure it looks like what we expect? At least that it's an array that can be walked through and maybe that each element could validly be converted to a `URL` type (ideally they would be stored in the array as a `URL` type but maybe that's too hard for compatibility reasons)